atomic_xadd::<f64> fails with rustc, succeeds with miri

Consider the following program:

```rust
// test.rs

#![feature(core_intrinsics)]

pub fn main() {
    let mut z: f64 = 1.0;
    unsafe {
        ::std::intrinsics::atomic_xadd(&mut z, 2.0);
    }
}
```

When I try to run `test.rs` with `rustc`, I get a monomorphization error:

```
$ rustc test.rs
error[E0511]: invalid monomorphization of `atomic_xadd` intrinsic: expected basic integer type, found `f64`
 --> test.rs:8:9
  |
8 |         ::std::intrinsics::atomic_xadd(&mut z, 2.0);
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to previous error(s)
```

However, when I run `test.rs` with `miri`, I expect to see a similar error, but instead the program finishes successfully:

```
$ cargo run --bin miri -- test.rs 
    Finished dev [unoptimized + debuginfo] target(s) in 0.0 secs
     Running `target/debug/miri test.rs`
```


A possible solution might be to add some typechecking in [intrinsic.rs](https://github.com/solson/miri/blob/14848b3badb64bf0c9eab43eb87ce07d61f8a4d2/src/terminator/intrinsic.rs#L118) to match the logic in [rustc_trans::intrinsic](https://github.com/rust-lang/rust/blob/0b17b4c0846d0e01544e90bb2aa2776576e825d1/src/librustc_trans/intrinsic.rs#L414-L417).

It would be even better if rustc could somehow catch these errors during typechecking, so that miri could inherit the error handling.

--- a/src/intrinsic.rs
+++ b/src/intrinsic.rs
@@ -126,6 +126,9 @@ impl<'a, 'mir, 'tcx> EvalContextExt<'tcx> for super::MiriEvalContext<'a, 'mir, '
             "atomic_xsub_acqrel" |
             "atomic_xsub_relaxed" => {
                 let ptr = self.ref_to_mplace(self.read_value(args[0])?)?;
+                if !ptr.layout.ty.is_integral() {
+                    return err!(Unimplemented(format!("Atomic arithmetic operations only work on integer types")));
+                }
                 let rhs = self.read_value(args[1])?;
                 let old = self.read_value(ptr.into())?;
                 self.write_value(*old, dest)?; // old value is returned

--- a/tests/compile-fail/atomic_non_integer_arithmetic.rs
+++ b/tests/compile-fail/atomic_non_integer_arithmetic.rs
@@ -1,0 +1,9 @@
+#![feature(core_intrinsics)]
+
+pub fn main() {
+    let mut z: f64 = 1.0;
+    unsafe {
+        ::std::intrinsics::atomic_xadd(&mut z, 2.0);
+        //~^ ERROR: Atomic arithmetic operations only work on integer types
+    }
+}

